source/simulator/solver_schemes.cc

```diff
@@ -0,0 +1,744 @@
+    // all fields, so that we use the same point in time for every field when solving
+    for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+      {
+        current_linearization_point.block(introspection.block_indices.compositional_fields[c])
+          = solution.block(introspection.block_indices.compositional_fields[c]);
+
+        if ((initial_residual != NULL) && (*initial_residual)[c] > 0)
+          current_residual[c] /= (*initial_residual)[c];
+        else
+          current_residual[c] = 0.0;
+      }
+
+    return current_residual;
+  }
+
+
+
+  template <int dim>
+  double Simulator<dim>::assemble_and_solve_stokes (const bool compute_initial_residual,
+                                                    double *initial_residual)
+  {
+    assemble_stokes_system ();
+    build_stokes_preconditioner();
+
+    if (compute_initial_residual)
+      {
+        Assert(initial_residual != NULL, ExcNotImplemented());
+        *initial_residual = compute_initial_stokes_residual();
+      }
+
+    const double current_residual = solve_stokes();
+
+    current_linearization_point.block(introspection.block_indices.velocities)
+      = solution.block(introspection.block_indices.velocities);
+
+    if (introspection.block_indices.velocities != introspection.block_indices.pressure)
+      current_linearization_point.block(introspection.block_indices.pressure)
+        = solution.block(introspection.block_indices.pressure);
+
+    if (parameters.include_melt_transport)
+      {
+        // Note that the compaction pressure is in the fluid pressure block
+        // and will therefore be updated as well.
+        const unsigned int fluid_velocity_block = introspection.variable("fluid velocity").block_index;
+        const unsigned int fluid_pressure_block = introspection.variable("fluid pressure").block_index;
+        current_linearization_point.block(fluid_velocity_block) = solution(fluid_velocity_block);
+        current_linearization_point.block(fluid_pressure_block) = solution.block(fluid_pressure_block);
+      }
+
+    if ((initial_residual != NULL) && (*initial_residual > 0))
+      return current_residual / *initial_residual;
+
+    return 0.0;
+  }
+
+
+
+  template <int dim>
+  void Simulator<dim>::solve_IMPES ()
+  {
+    assemble_and_solve_temperature();
+    assemble_and_solve_composition();
+
+    // the Stokes matrix depends on the viscosity. if the viscosity
+    // depends on other solution variables, then after we need to
+    // update the Stokes matrix in every time step and so need to set
+    // the following flag. if we change the Stokes matrix we also
+    // need to update the Stokes preconditioner.
+    if (stokes_matrix_depends_on_solution() == true)
+      rebuild_stokes_matrix = rebuild_stokes_preconditioner = true;
+    assemble_and_solve_stokes();
+
+    if (parameters.run_postprocessors_on_nonlinear_iterations)
+      postprocess ();
```

Comment:
Is this function (`solve_IMPES`) called in a nonlinear loop? If not, shouldn't we always call this?

(I guess my question must be dumb because surely this would break a million output files otherwise...)

Reply:
This function is not called inside a nonlinear loop, but the execution of postprocessor in the `run` function is disabled if we run them on every nonlinear iteration. The responsibility of calling them is transferred to the solver scheme in that case. Thus, even for solver schemes that do not have a nonlinear loop, they are responsible for calling the postprocessors if they should be called on every nonlinear iteration. In the end, the postprocessors are still only called once for the IMPES scheme (only here, instead of the main loop).